includes/class-wpcom-legacy-redirector-ui.php

```diff
@@ -0,0 +1,429 @@
+						), $sendback
+					) );
+					exit();
+				}
+				// Check if $redirect is valid.
+				if ( wp_validate_redirect( $redirect, false ) && 404 !== $status || 'valid' === $redirect ) {
+					wp_safe_redirect( add_query_arg(
+						array(
+							'validate'  => 'valid',
+							'ids'       => $post->ID,
+						), $sendback
+					) );
+					exit();
+				}
+			}
+		}
+	}
+	public function generate_page_html() {
+		if ( ! current_user_can( 'manage_redirects' ) ) {
+			return;
+		}
+
+		$errors   = array();
+		$messages = array();
+		if ( class_exists( 'WPCOM_Legacy_Redirector' ) ) {
+			if ( isset( $_POST['from_url'] ) && isset( $_POST['redirect_to'] ) ) {
+				if (
+					! isset( $_POST['redirect_nonce_field'] )
+					|| ! wp_verify_nonce( $_POST['redirect_nonce_field'] )
+				) {
+					$errors[] = array(
+						'label'   => __( 'Error', 'wpcom-legacy-redirector' ),
+						'message' => __( 'Sorry, your nonce did not verify.', 'wpcom-legacy-redirector' ),
+					);
+				} else {
+					$from_url	= sanitize_text_field( $_POST['from_url'] );
+					$redirect_to = sanitize_text_field( $_POST['redirect_to'] );
+
+					// Check if $from_url is not a 404.  Only 404 links are redirected in this plugin.
+					$status = $this->check_if_404( home_url() . $from_url );
+					if ( 404 !== $status ) {
+						$from_url = '';
+					}
+
+					if ( WPCOM_Legacy_Redirector::validate( $from_url, $redirect_to ) ) {
+						$output = WPCOM_Legacy_Redirector::insert_legacy_redirect( $from_url, $redirect_to, false );
+						if ( true === $output ) {
+							$link	   = '<a href="' . esc_url( $from_url ) . '" target="_blank">' . esc_url( $from_url ) . '</a>';
+							$messages[] = __( 'The redirect was added successfully. Check Redirect: ', 'wpcom-legacy-redirector' ) . $link;
+						} else {
+							if ( false === $output ) {
+								$errors[] = array(
+									'label'   => __( 'Error', 'wpcom-legacy-redirector' ),
+									'message' => __( 'Redirect could not be saved. Contact administrator and check permissions.', 'wpcom-legacy-redirector' ),
+								);
+							} elseif ( is_wp_error( $output ) ) {
+								foreach ( $output->get_error_messages() as $error ) {
+									$errors[] = array(
+										'label'   => __( 'Error', 'wpcom-legacy-redirector' ),
+										'message' => $error,
+									);
+								}
+							}
+						}
+					} else {
+						if ( $from_url === '' ) {
+							$errors[] = array(
+								'label'   => __( 'Error', 'wpcom-legacy-redirector' ),
+								'message' => __( 'Redirects need to be from URLs that have a 404 status.', 'wpcom-legacy-redirector' ),
+							);
+							$output = false;
+						} else {
+							$errors[] = array(
+								'label'   => __( 'Error', 'wpcom-legacy-redirector' ),
+								'message' => __( 'Check the values you are using to save the redirect. All fields are required. "Redirect From" and "Redirect To" should not match.', 'wpcom-legacy-redirector' ),
+							);
+						}
+					}
+				}
+			}
+		} else {
```

Comment:
For conditionals like this, if we inverted the first conditional we could have it perhaps return early? or do something to remove the extra indentation.